crates/noirc_evaluator/src/ssa/acir_gen/constraints.rs

```diff
@@ -127,7 +126,13 @@ pub(crate) fn subtract(a: &Expression, k: FieldElement, b: &Expression) -> Expre
 // TODO also check why we are doing all of this complicated logic with i1 and i2
 // TODO in either case, we can put this in ACIR, if its useful
 pub(crate) fn add(a: &Expression, k: FieldElement, b: &Expression) -> Expression {
-    let mut output = Expression::default();
+    if a.is_const() {
+        return (b * &k) + &a.q_c;
+    } else if b.is_const() {
+        return a.clone() + &(k * b.q_c);
```

Comment:
I would avoid the + and just clone a and then a_clone.q_c += k*b.q_c and return a_clone

Reply:
My point was to not use the + which is not efficient (when adding two expressions), but in fact I see that there is a + for expression + field which is doing what we want.
However it is definitively less readable as the code is scattered through several implicit function and you don't know who is doing what and who is copying/cloning what. The fact that I did not see is proving my point.
But if that code is doing clone + only update the constant term, then I am fine with this.
Can somebody confirm this?

Reply:
Yes, the `Expression + &FieldElement` operation should be efficient as we're moving the vectors into the resulting `Expression` rather than rebuilding them as we would do for `Expression + Expression`.

Part of the readability issues may be from the fact that the definitions of these operators are all the way down in `acir` compared to these functions being up here. We've then got a weird situation with parallel implementations of some functions. My preference is for `add`, `subtract` and `mul` to be pushed down into `acir`.

Reply:
Thank you for your answer.
Moving functions away is not going to make this more readable to me.